src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -1072,42 +1075,39 @@ def dbapi(
             sql_create_temp_stage = f"create {get_temp_type_for_object(self._session._use_scoped_temp_objects, True)} stage if not exists {snowflake_stage_name}"
             self._session._run_query(sql_create_temp_stage, is_ddl_on_temp_object=True)
 
-            with ProcessPoolExecutor(max_workers=max_workers) as executor:
-                futures = [
-                    executor.submit(
-                        task_fetch_from_data_source,
-                        create_connection,
-                        query,
-                        raw_schema,
-                        i,
-                        tmp_dir,
-                    )
-                    for i, query in enumerate(partitioned_queries)
-                ]
-
-                completed_futures = wait(futures, return_when=ALL_COMPLETED)
-            files = []
-            for f in completed_futures.done:
-                if isinstance(f.result(), Exception):
-                    raise f.result()
-                else:
-                    files.append(f.result())
-            with ThreadPoolExecutor(max_workers=max_workers) as thread_executor:
-                futures = [
-                    thread_executor.submit(
-                        self.upload_and_copy_into_table,
-                        f,
-                        snowflake_stage_name,
-                        snowflake_table_name,
-                        "abort_statement",
+            with ProcessPoolExecutor(max_workers=max_workers) as process_executor:
+                with ThreadPoolExecutor(max_workers=max_workers) as thread_executor:
+                    thread_pool_futures = []
+                    process_pool_futures = [
+                        process_executor.submit(
+                            task_fetch_from_data_source_with_retry,
+                            create_connection,
+                            query,
+                            raw_schema,
+                            i,
+                            tmp_dir,
+                        )
+                        for i, query in enumerate(partitioned_queries)
+                    ]
+                    for future in as_completed(process_pool_futures):
+                        if isinstance(future.result(), Exception):
+                            raise future.result()
+                        else:
+                            thread_pool_futures.append(
+                                thread_executor.submit(
+                                    self.upload_and_copy_into_table_with_rename_with_retry,
+                                    future.result(),
+                                    snowflake_stage_name,
+                                    snowflake_table_name,
+                                    "abort_statement",
+                                )
+                            )
+                    completed_futures = wait(
+                        thread_pool_futures, return_when=ALL_COMPLETED
                     )
```

Comment:
same here, I'm wondering if we can introduce a broadcast mechanism here, if one job failed, the rest needs to be cancelled.